torch/csrc/jit/codegen/cuda/lower2device.cpp

```diff
@@ -172,6 +172,98 @@ std::unordered_map<Val*, Val*> getSimplificationMap(Fusion* fusion) {
   return extent_to_min_input_id_extent;
 }
 
+bool shouldFuseBroadCast(const BroadcastOp* broadcast) {
+  auto broadcast_inp = broadcast->in()->as<TensorView>();
+
+  // skip if the broadcast directly uses fusion input,
+  //  meaning there's no reduction to be fused with.
+  if (broadcast_inp->isFusionInput()) {
+    return false;
+  }
+
+  // Check that the producer of the input is a warp reduction,
+  //  which could fuse with block broadcast on threadIdx.x
+  if (auto reduction_op =
+          dynamic_cast<ReductionOp*>(broadcast_inp->definition())) {
+    if (!ir_utils::getMaybeWarpReductionDim(reduction_op).has_value()) {
+      return false;
+    }
+  } else {
+    return false;
+  }
+
+  auto broadcast_out = broadcast->out()->as<TensorView>();
+
+  // Only support fusing register to register operations
+  //  at this stage.
+  if (broadcast_inp->getMemoryType() != MemoryType::Local ||
+      broadcast_out->getMemoryType() != MemoryType::Local) {
+    return false;
+  }
+
+  // Check that the broadcast operation is only either serial
+  //  or across threadIdx.x of the block.
+  const auto parallel_domains =
+      GpuLower::current()->threadPredMap().getParallelBroadcastDomains(
+          broadcast_out);
+
+  // Won't change anything if this broadcast doesn't
+  //  parallelize on TIDX.
+  if (!parallel_domains.get(ParallelType::TIDx)) {
+    return false;
+  }
+
+  // Locate the broadcast axes and
+  //  check if the broadcast has other thread parallelization
+  int first_broadcast_pos = -1;
+  for (size_t i = 0; i < broadcast_out->nDims(); i++) {
+    if (broadcast_out->axis(i)->isBroadcast()) {
+      // Record the first broadcast id on the tv
+      if (first_broadcast_pos == -1) {
+        first_broadcast_pos = i;
+      }
+      // Check if the broadcast is parallelized on any other
+      //  dimension. Wouldn't be able to fuse if so.
+      if (broadcast_out->axis(i)->isParallelized() &&
+          broadcast_out->axis(i)->getParallelType() != ParallelType::TIDx &&
+          parallel_domains.get(broadcast_out->axis(i)->getParallelType())) {
+        return false;
+      }
+    }
+  }
+
+  // Find the warp reduction position of the producer reduction
+  int reduction_pos = -1;
+  for (size_t i = 0; i < broadcast_inp->nDims(); i++) {
+    if (broadcast_inp->axis(i)->isReduction() &&
```

Comment:
Can you explicitly allow any reduction axis with a static size 1? This can happen if we split a dimension for unswitch but it's just size 1 so it's just like a trivial reduction in this case.

Reply:
Updated thanks.